dbcon/mysql/ha_mcs_pushdown.cpp

```diff
@@ -830,31 +801,55 @@ create_columnstore_select_handler(THD* thd, SELECT_LEX* select_lex)
             item_check(item, &unsupported_feature);
             if (unsupported_feature)
             {
-                break;
+                return handler;
             }
         }
     }
 
-    if (!unsupported_feature)
+    // We apply dedicated rewrites from MDB here so MDB's data structures
+    // becomes dirty and CS has to raise an error in case of any problem
+    // or unsupported feature.
+    handler= new ha_columnstore_select_handler(thd, select_lex);
+    JOIN *join= select_lex->join;
     {
-        handler= new ha_columnstore_select_handler(thd, select_lex);
-        mcs_handler_info mhi= mcs_handler_info(reinterpret_cast<void*>(handler), SELECT);
-        // handler::table is the place for the result set
-        int rc= 0;
-        // Skip execution for EXPLAIN queries
-        if (!thd->lex->describe)
+        disable_indices_for_CEJ(thd);
+
+        if (select_lex->handle_derived(thd->lex, DT_MERGE))
         {
-            rc= ha_cs_impl_pushdown_init(&mhi, handler->table);
+            unsupported_feature = true;
+            handler->err_msg.assign("create_columnstore_select_handler(): Internal error happens in SL::handle_derived()");
         }
 
-        // Return SH even if init fails b/c CS changed SELECT_LEX structures
-        // with simplify_joins_()
-        if (rc)
+        COND *conds = nullptr;
+        if (!unsupported_feature)
+        {
+            conds = simplify_joins_mcs(join, select_lex->join_list,
+                join->conds, TRUE, FALSE);
+        }
+        // MCOL-3747 IN-TO-EXISTS rewrite inside MDB didn't add
+        // an equi-JOIN condition.
+        if (!unsupported_feature
+            && optimize_unflattened_subqueries_mcs(select_lex))
+        {
             unsupported_feature = true;
-        return handler;
+            handler->err_msg.assign("create_columnstore_select_handler: Internal error happens in optimize_unflattened_subqueries_mcs()");
```

Comment:
Instead of happens, occurred would be a better word here.

Reply:
To be honest it at this point the error happens for real not occured :)